### ISSUE TITLE
Stop() seems to be hanging

### ISSUE THREAD

**Issue by kall2sollies** [Marfusios/websocket-client#69]
Hi,

I have some trouble managing the channel close on client request.
I'm in an ASP.Net Core / .NetCore 5 environment.

I wrote a wrapper as an injected service with Singleton scope around your library which you can find below.
This service is called on background using Hangfire.

When I `Ctrl+C` the console server, I managed to tell the Hangfire job to call the stop method of my websocket service. `Stop()` fires the exit event, so that the second part of the `Listen()` method (where it stops) is executed.
The end of the `Listen()` method is reached, but later on, Hangfire throws the exception when a job was not gracefully stopped.
I think there is an issue with disposing the client, and / or subscriptions to events.

````csharp
private WebsocketClient _client;
private ManualResetEvent _exitEvent;
public bool Connected;

public void Start()
{
	// Logging
	var identifier = $"[ {nameof(StandardSocketService)}.{nameof(Start)} ]";
	_logger.LogInformation($"{identifier} Starting connexion to WebSocket {_endpoint}");

	// Connect and listen
	_exitEvent = new ManualResetEvent(false);
	var url = new Uri(_endpoint);

	// callbacks
	_client = new WebsocketClient(url) { ReconnectTimeout = TimeSpan.FromMilliseconds(ReconnexionTimeout) };
	_client.DisconnectionHappened.Subscribe(info =>
	{
		Connected = false;
		_logger.LogInformation($"{identifier} Disconnection happened, type: {info.Type}");
	});
	_client.ReconnectionHappened.Subscribe(info =>
	{
		_logger.LogInformation($"{identifier} Reconnection happened, type: {info.Type}");
	});
	_client.MessageReceived.Subscribe(msg =>
	{
		_logger.LogInformation($"{identifier} Message received: {msg}");
	});
	// Actual start
	_client.Start();
	Connected = true;
	_logger.LogInformation($"{identifier} Started connexion to WebSocket {_endpoint}");
}

public void Listen()
{
	if (!Connected) Start();
	// Logging
	var identifier = $"[ {nameof(StandardSocketService)}.{nameof(Listen)} ]";
	_logger.LogInformation($"{identifier} Listening to WebSocket {_endpoint}");
	_exitEvent.WaitOne(); // block here until receive _exitEvent.Set();

	// We reach here when shutdown of application is required
	_client.IsReconnectionEnabled = false;
	_client.Stop(WebSocketCloseStatus.Empty, string.Empty);
	Connected = false;
	_logger.LogWarning($"{identifier} Closed connexion to WebSocket {_endpoint}");
}

public void Send(string data)
{
	if (!Connected) Listen();
	_client.Send(data);
}

public void Stop() => _exitEvent.Set();
````

**Comment by Marfusios** [comment]
Hello @kall2sollies, 

what exception does Hangfire throw? 

`Stop` is called when the application is closing down or `Stop` and `Start` could be called multiple times per app lifecycle? 

In the case of the former, you should definitely await stop method (`await _client.Stop(WebSocketCloseStatus.Empty, string.Empty);`) and also call `_client.Dispose()` at the end.

**Comment by kall2sollies** [comment]
If possible, I'd prefer keep my implementation in a sync environment, so that's why I haven't awaited the Stop() method. But I tried to call `_client.Stop().Wait()` (I saw that in one of your exemples) and it wasn't better.

The Hangfire exception is the famous one you get when a background job doesn't respond in time to the CancellationToken.

Here is the excerpt of my log file:

````
[2021-02-25 11:26:28.531][Hangfire.Server.BackgroundServerProcess][WRN] Server winterfell:16520:e415fc81 caught shutdown signal...
[2021-02-25 11:26:28.532][Hangfire.Server.BackgroundServerProcess][WRN] Server winterfell:16520:e415fc81 stopped non-gracefully due to Worker. Outstanding work on those dispatchers could be aborted, and there can be delays in background processing. This server instance will be incorrectly shown as active for a while. To avoid non-graceful shutdowns, investigate what prevents from stopping gracefully and add CancellationToken support for those methods.
````

**Comment by kall2sollies** [comment]
I finally managed to `async`-ify my code and there is no issue anymore, shutdowns are clean now. And thanks to that, I no longer need the `_exitEvent.WaitOne()`.

**Comment by kall2sollies** [comment]
For the record, here is my full service implementation:

````csharp
public class StandardSocketService : IStandardSocketService
{
	private readonly ILogger<StandardSocketService> _logger;
	private readonly string _endpoint;
	private WebsocketClient _client;
	public bool Connected;
	private const int ReconnexionTimeout = 30 * 1000;

	public StandardSocketService(
		IOptions<FranceFoodSettings> options,
		ILogger<StandardSocketService> logger)
	{
		var settings = options.Value;
		_logger = logger;

		_endpoint = "wss://echo.websocket.org";
	}

	public async Task Start()
	{
		// Logging
		var identifier = $"{nameof(StandardSocketService)}.{nameof(Start)}";
		_logger.LogInformation($"{identifier} Starting connexion to WebSocket {_endpoint}");

		// Connect and listen
		var url = new Uri(_endpoint);

		// callbacks
		_client = new WebsocketClient(url) { ReconnectTimeout = TimeSpan.FromMilliseconds(ReconnexionTimeout) };
		_client.DisconnectionHappened.Subscribe(info =>
		{
			Connected = false;
			_logger.LogDebug($"[ {identifier}.DisconnectionHappened ] Disconnection happened, type: {info.Type}");
		});
		_client.ReconnectionHappened.Subscribe(info =>
		{
			_logger.LogDebug($"[ {identifier}.ReconnectionHappened ] Reconnection happened, type: {info.Type}");
		});
		_client.MessageReceived.Subscribe(msg =>
		{
			_logger.LogInformation($"[ {identifier}.MessageReceived ] Message received: {msg}");
		});
		// Actual start
		await _client.Start();
		Connected = true;
		_logger.LogInformation($"{identifier} Started connexion to WebSocket {_endpoint}");
	}

	public async Task Listen()
	{
		if (!Connected) await Start();
		// Logging
		var identifier = $"[ {nameof(StandardSocketService)}.{nameof(Listen)} ]";
		_logger.LogInformation($"{identifier} Listening to WebSocket {_endpoint}");
	}

	public async Task Send(string data)
	{
		if (!Connected) await Listen();
		_client.Send(data);
	}

	public async Task Stop()
	{
		// Logging
		var identifier = $"[ {nameof(StandardSocketService)}.{nameof(Stop)} ]";
		_client.IsReconnectionEnabled = false;
		await _client.Stop(WebSocketCloseStatus.Empty, string.Empty);
		Connected = false;
		_logger.LogWarning($"{identifier} Closed connexion to WebSocket {_endpoint}");
	}
}
````

And the Hangfire job that runs it all:

````csharp
public override async Task RunAsyncImplementation()
{
	// Start the service in a new thread
	var thread = new Thread(async () => await _socketService.Listen());
	thread.Start();
	Logger.LogInformation($"[ {JobId} ] Websocket listener has starded");
	// Listen to graceful stop demands
	while (true)
	{
		if (IsCancelRequested) // this is how I propagate JobCancellationTokens
		{
			await _socketService.Stop();
			break;
		}
		await Task.Delay(500);
	}
}
````

**Comment by Marfusios** [comment]
Great to hear that, async/await top-down is definitely a correct approach. 

Btw, a few tips. 
1. do not use old System.Threading when not needed. `Thread` is a very low-level concept, it is much better to use `Tasks` - it shares threads in the background, saves resources, etc. You can call `Task.Run()` or `Task.Factory.StartNew()`. 
2. avoid that kind of `while(true)` cycles, it unnecessarily consumes CPU

In your use-case, you can do something like: 

```csharp
public override async Task RunAsyncImplementation()
{
        await _socketService.Listen();
	Logger.LogInformation($"[ {JobId} ] Websocket listener has started");
	
        finishedEvent = new AsyncManualResetEvent(set: false);
        _cancellationToken.Register(() => finishedEvent.Set());

        await _finishedEvent.WaitAsync();
        await _socketService.Stop();
}
```

Source code for `AsyncManualResetEvent` is for example here: https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncManualResetEvent.cs

**Comment by kall2sollies** [comment]
Thanks a lot for the tips, I'll dig further about `AsyncManualResetEvent`. As a pure Web backend developer, dealing with threads is way out of my comfort zone !
And thanks for the `Task.Factory.StartNew()`, I remembered the `Thread.Start()` was almost deprecated, but couldn't remember of the former.
Anyway, since then, I gave up the Thread thing for a simple `await` exactly the same way you mentioned. I'll just have to replace the infinite loop given your hints.
